File: android/app/src/main/kotlin/com/tampagamingguild/tggmobile/BeaconBackgroundRunner.kt
package com.tampagamingguild.tggmobile

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.FlutterCallbackInformation

/**
 * Runs the Dart side of a background beacon detection.
 *
 * The check-in itself lives in Dart, not here: it needs the stored refresh
 * token, the token-rotation logic, the once-per-day stamp and the same
 * api/checkins.php call the foreground path uses. Reimplementing all of that
 * in Kotlin would mean two copies to keep in step, so instead this spins up a
 * headless FlutterEngine and calls into the app's own code.
 *
 * The Dart entrypoint to run is registered by the app while it's in the
 * foreground (see BeaconBackgroundChannel) and kept in SharedPreferences,
 * since by the time a broadcast arrives there may be no app process at all.
 */
object BeaconBackgroundRunner {
    private const val TAG = "BeaconBackgroundRunner"
    const val BACKGROUND_CHANNEL = "com.tampagamingguild.tggmobile/beacon_background"

    /** Ceiling on how long Dart gets before we tear the engine down anyway. */
    private const val TIMEOUT_MS = 25_000L

    private var engine: FlutterEngine? = null

    fun dispatch(context: Context, onComplete: () -> Unit) {
        // FlutterEngine must be created and driven from the main thread;
        // onReceive may not be on it.
        Handler(Looper.getMainLooper()).post { start(context, onComplete) }
    }

    private fun start(context: Context, onComplete: () -> Unit) {
        if (engine != null) {
            Log.i(TAG, "Background engine already running; ignoring duplicate detection")
            onComplete()
            return
        }

        val handle = BeaconBackgroundChannel.callbackHandle(context)
        if (handle == 0L) {
            Log.w(TAG, "No Dart callback registered; cannot run background check-in")
            onComplete()
            return
        }

        // Must precede any FlutterJNI call. Resolving a callback handle is a
        // native lookup, so doing it first throws UnsatisfiedLinkError in the
        // case that matters most: a process cold started purely to deliver
        // this broadcast, where nothing has loaded libflutter yet. With the
        // app merely backgrounded the library is already up and the wrong
        // order appears to work.
        val loader = io.flutter.FlutterInjector.instance().flutterLoader()
        if (!loader.initialized()) {
            loader.startInitialization(context)
        }
        loader.ensureInitializationComplete(context, null)

        val callback = FlutterCallbackInformation.lookupCallbackInformation(handle)
        if (callback == null) {
            Log.w(TAG, "Registered Dart callback could not be resolved")
            onComplete()
            return
        }

        val created = FlutterEngine(context)
        engine = created

        // The background isolate needs the control channel too, not just the
        // main engine's copy: after a successful check-in it cancels the
        // registered scan, and without this that call lands on a channel with
        // no handler and is silently swallowed, leaving the OS waking us for
        // every further sighting all visit.
        BeaconBackgroundChannel.register(context, created.dartExecutor.binaryMessenger)

        var finished = false
        val timeout = Handler(Looper.getMainLooper())
        lateinit var finish: (String) -> Unit

        finish = { reason ->
            if (!finished) {
                finished = true
                timeout.removeCallbacksAndMessages(null)
                Log.i(TAG, "Background check-in finished ($reason)")
                created.destroy()
                engine = null
                onComplete()
            }
        }

        val channel = MethodChannel(created.dartExecutor.binaryMessenger, BACKGROUND_CHANNEL)
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                // Dart has wired up its handler and is ready to be told what
                // happened; anything sent before this would be dropped.
                "backgroundHandlerReady" -> {
                    result.success(null)
                    channel.invokeMethod("onBeaconDetected", null)
                }
                "backgroundHandlerDone" -> {
                    result.success(null)
                    finish("reported done")
                }
                else -> result.notImplemented()
            }
        }

        timeout.postDelayed({ finish("timed out") }, TIMEOUT_MS)

        created.dartExecutor.executeDartCallback(
            DartExecutor.DartCallback(context.assets, loader.findAppBundlePath(), callback)
        )
    }
}

File: ios/Runner/BeaconBackgroundRunner.swift
import Flutter
import UIKit

/// Runs the Dart side of a background beacon detection -- iOS counterpart of
/// BeaconBackgroundRunner.kt.
///
/// The check-in itself lives in Dart, not here: it needs the stored refresh
/// token, the token-rotation logic, the once-per-day stamp and the same
/// api/checkins.php call the foreground path uses. Reimplementing all of
/// that here would mean two copies to keep in step, so instead this spins
/// up a headless FlutterEngine and calls into the app's own code, the same
/// way the Android runner does.
///
/// The Dart entrypoint to run is registered by the app while it's in the
/// foreground (see BeaconBackgroundChannel) and kept in UserDefaults, since
/// by the time a region entry arrives there may be no app process at all.
enum BeaconBackgroundRunner {
    // Must match backgroundChannel in beacon_background.dart exactly.
    static let backgroundChannelName = "com.tampagamingguild.tggmobile/beacon_background"

    /// Ceiling on how long Dart gets before the engine is torn down anyway,
    /// with margin below iOS's own background-task expiration.
    private static let timeoutSeconds: TimeInterval = 20

    private static var engine: FlutterEngine?

    static func dispatch(completion: @escaping () -> Void) {
        // FlutterEngine must be created and driven from the main thread;
        // the CLLocationManagerDelegate callback that leads here may not be.
        DispatchQueue.main.async { start(completion: completion) }
    }

    private static func start(completion: @escaping () -> Void) {
        if engine != nil {
            // A detection is already being handled; ignore the duplicate,
            // mirroring the Android runner's same guard.
            completion()
            return
        }

        let handle = BeaconBackgroundChannel.callbackHandle
        guard handle != 0, let info = FlutterCallbackCache.lookupCallbackInformation(handle) else {
            completion()
            return
        }

        var bgTask: UIBackgroundTaskIdentifier = .invalid
        var finished = false

        func finish(_ reason: String) {
            guard !finished else { return }
            finished = true
            engine?.destroyContext()
            engine = nil
            if bgTask != .invalid {
                UIApplication.shared.endBackgroundTask(bgTask)
                bgTask = .invalid
            }
            completion()
        }

        bgTask = UIApplication.shared.beginBackgroundTask(withName: "BeaconCheckIn") {
            finish("expired")
        }

        let created = FlutterEngine(name: "BeaconBackgroundEngine")
        guard created.run(withEntrypoint: info.callbackName, libraryURI: info.callbackLibraryPath) else {
            NSLog("BeaconBackgroundRunner: failed to start Dart callback")
            finish("failed to start")
            return
        }

        // Unlike Android, where a freshly created FlutterEngine registers
        // plugins for itself, iOS requires this explicitly -- skip it and
        // NotificationService/AutoCheckinPreference's channels
        // (flutter_local_notifications, flutter_secure_storage) have no
        // native handler inside this engine and silently fail inside
        // runCheckIn().
        GeneratedPluginRegistrant.register(with: created)
        engine = created

        // The headless isolate needs its own copy of the control channel
        // too: after a successful check-in it cancels the registered
        // region, and without this that call lands on a channel with no
        // handler and is silently swallowed, leaving the OS waking the app
        // for every further sighting the rest of the day.
        BeaconBackgroundChannel.register(with: created.binaryMessenger)

        let channel = FlutterMethodChannel(name: backgroundChannelName, binaryMessenger: created.binaryMessenger)
        channel.setMethodCallHandler { call, result in
            switch call.method {
            // Dart has wired up its handler and is ready to be told what
            // happened; anything sent before this would be dropped.
            case "backgroundHandlerReady":
                result(nil)
                channel.invokeMethod("onBeaconDetected", arguments: nil)
            case "backgroundHandlerDone":
                result(nil)
                finish("reported done")
            default:
                result(FlutterMethodNotImplemented)
            }
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + timeoutSeconds) {
            finish("timed out")
        }
    }
}
